src/utils/clone-deep.js

```diff
@@ -1,5 +1,44 @@
-function cloneDeep(obj) {
-  return JSON.parse(JSON.stringify(obj));
+function isObject(value) {
+  return Object.prototype.toString.call(value) === '[object Object]';
+}
+
+function isArray(value) {
+  return Array.isArray(value);
+}
+
+function isFunction(value) {
+  return typeof value === 'function';
+}
+
+function isDate(value) {
+  return value instanceof Date;
+}
+
+function isPrimitiveType(value) {
+  return ['string', 'number', 'boolean', 'undefined'].includes(typeof value);
+}
+
+function cloneDeep(value) {
+  if (isPrimitiveType(value) || isFunction(value) || value === null) {
+    return value;
+  }
+
+  if (isDate(value)) {
+    return new Date(value.getTime());
+  }
+
+  if (isArray(value)) {
+    return value.map(item => cloneDeep(item));
+  }
+
+  if (isObject(value)) {
+    return Object.entries(value).reduce((clone, entry) => {
+      clone[entry[0]] = cloneDeep(entry[1]);
+      return clone;
+    }, {});
+  }
+
+  return value;
```

Comment:
It doesn't work with `BigInt` and `Symbol` types but we don't use them. Should we add `BigInt`, just in case (I think there will never be any symbols in pure data structures)?

Reply:
I considered `Symbol` but also thought it's not the case for us here.
I did not think of `BigInt` since I never used it before, but I can surely add it.

Reply:
It won't hurt

Reply:
Done!
Actually we could simplify the method to the following, since the check for `isObject` is strict and would not include `null`, `Date`, `Array` (so no danger if the `if` statements would be switched around):

```
function cloneDeep(value) {
  if (isDate(value)) {
    return new Date(value.getTime());
  }

  if (isArray(value)) {
    return value.map(item => cloneDeep(item));
  }

  if (isObject(value)) {
    return Object.entries(value).reduce((clone, entry) => {
      clone[entry[0]] = cloneDeep(entry[1]);
      return clone;
    }, {});
  }

  return value;
}
```

Or would we rather specifically see the `isPrimitiveType` and `isFunction` checks?
Thoughts?